### DISCUSSION TITLE
Trying to send file buffer using libp2p example code

### DISCUSSION THREAD

**Discussion by NoorahSmith** [libp2p/rust-libp2p#4372] [General]
I am trying to send file from one terminal to other terminal by using the filesharing example. libp2p 0.52.2 version but the code keeps on running and doesn't give error on receiver side. both nodes are on the same computer with different seed.  In receiver i don't get the provider is empty but both these nodes keep on running

Sender code
 
```
let secretkeyseed: std::option::Option<u8>=Some(1);//get seed

    let (mut network_client, mut network_events, network_event_loop) =
        network::new(secretkeyseed).await?;

    // Spawn the network task for it to run in the background.
    spawn(network_event_loop.run());
    
    //network_client.get_providers("public1.txt".to_string()).await;
    let mut name="public1.txt".to_string();
    network_client.start_providing(name.clone()).await;
    //loop {
            
                    match network_events.next().await {
                        // Reply with the content of the file on incoming requests.
                        Some(network::Event::InboundRequest { request, channel }) => {
                            if request == name {
                                network_client
                                    .respond_file(std::fs::read("/home/rusty/code/datafile.bin".to_string())?, channel)
                                    .await;
                            }
                        }
                        e => todo!("{:?}", e),
                    }
      //          }
```

Receiver Code

```
 let secretkeyseed: std::option::Option<u8>=Some(10);

    let (mut network_client, mut network_events, network_event_loop) =
        network::new(secretkeyseed).await?;
        

    // Spawn the network task for it to run in the background.
    spawn(network_event_loop.run());
    let mut name="public1.txt".to_string();
    network_client.get_providers(name.clone()).await;
    
    let providers = network_client.get_providers(name.clone()).await;
             if providers.is_empty() {
                return Err(format!("Could not find provider for file {name}.").into());
            }

            // Request the content of the file from each node.
            let requests = providers.into_iter().map(|p| {
                let mut network_client = network_client.clone();
                let name = name.clone();
                async move { network_client.request_file(p, name).await }.boxed()
            });

            // Await the requests, ignore the remaining once a single one succeeds.
            let file_content = futures::future::select_ok(requests)
                .await
                .map_err(|_| "None of the providers returned file.")?
                .0;
            println!("{:?}",file_content);

```

**Comment by dariusc93** [comment]
How are both nodes connected to each other?

**Reply by NoorahSmith** [reply]
Yes both nodes are on the same host . Listening on localhost with different seed

**Reply by dariusc93** [reply]
Are both connected to each other? Eg does your code contain something like https://github.com/libp2p/rust-libp2p/blob/master/examples/identify/src/main.rs#L59-L63 where you can supply the multiaddr in the command line (assuming its not fixed in the code)?

**Comment by NoorahSmith** [comment]
i am using 
network_client.start_listening("/ip4/0.0.0.0/tcp/0".parse()?).await.expect("Listening not to fail.");

**Reply by thomaseizinger** [reply]
Listening alone does not connect them. Are you issuing a dial? Do you see any events that say `ConnectionEstablished`?

Have a look at our examples, they all connect nodes to each other!

**Comment by NoorahSmith** [comment]
do i need to dial into the other node as well?
i thought that going through the providers and then requesting the file name would be sufficient . 

````
 let providers = network_client.get_providers(name.clone()).await;
             if providers.is_empty() {
                return Err(format!("Could not find provider for file {name}.").into());
            }

            // Request the content of the file from each node.
            let requests = providers.into_iter().map(|p| {
                let mut network_client = network_client.clone();
                let name = name.clone();
                async move { network_client.request_file(p, name).await }.boxed()
            });
```

**Comment by thomaseizinger** [comment]
> do i need to dial into the other node as well?
> i thought that going through the providers and then requesting the file name would be sufficient . 
> 
> ````
>  let providers = network_client.get_providers(name.clone()).await;
>              if providers.is_empty() {
>                 return Err(format!("Could not find provider for file {name}.").into());
>             }
> 
>             // Request the content of the file from each node.
>             let requests = providers.into_iter().map(|p| {
>                 let mut network_client = network_client.clone();
>                 let name = name.clone();
>                 async move { network_client.request_file(p, name).await }.boxed()
>             });
> ```

You seem to misunderstand what kademlia can and cannot do for you. If you haven't already, read/skim the spec to get an idea: https://github.com/libp2p/specs/tree/master/kad-dht. The actual kademlia paper is also worth a read.

Somewhere you need to connect to at least one node of your DHT. If you only have two nodes in total, you need to connect to the other one. If you only ever plan on having two nodes, you might want to reconsider your use of a DHT altogether.

**Comment by NoorahSmith** [comment]
Will look into kad-DHT
